Adds Mastercomfig support

BernardoGomesNegri opened the pull request:
The launcher is now capable of installing the optimization mod [Mastercomfig](https://mastercomfig.com/) and there is a new source for mods called "githubcollection". This also allows the user to switch between the different "versions" of mastercomfig
Most functions related to mods now take an optional parameter to specify the collection "version".
Here is a screenshot:
![image](https://user-images.githubusercontent.com/38192358/118323936-b904cd00-b4d7-11eb-9b70-f7c185b00f3c.png)

BernardoGomesNegri commented:
It says here that this pull request tries to change a workflow. It does not. This pull request does not change files under .github/workflows.

BernardoGomesNegri commented:
Now Installing Mastercomfig moves the autoexec.cfg and class cfgs to the user/ folder, as descibed in the [docs](https://docs.mastercomfig.com/en/latest/customization/custom_configs/). Uninstalling reverts the process.

rob5300 commented:
This looks great! I will get to reviewing this soon and we will also re check with mastercoms for including this. We got confirmation from her in the past but it was some time ago.

rob5300 commented on modules/mod_sources/collection_source.ts:
```diff
@@ -0,0 +1,15 @@
+//The same as the Mod-install_Source, but I had to create another, since TS does not let overriding abstract methods.
```
I don't think you need to create this duplicate class, i want to avoid this duplication.
You should be able to use the mod source base just fine, but if needed we can make changes to make it work.

BernardoGomesNegri replied:
This works, together with the array for all sources you mentioned.

rob5300 commented on renderer.js:
```diff
@@ -92,6 +94,27 @@ function OnClick_Mod(data) {
+        select.innerHTML = '';
+        //Populate the select
+        let defaultItem
+        data.items.forEach(element => {
+            let opt = document.createElement("option");
+            opt.value = element.itemname;
+            opt.innerHTML = element.displayname;
+            select.appendChild(opt);
+            if (element.default == true) {
+                opt.selected = true;
+            }
+        });
```
This is good! Will get @jota11 to improve the styling

rob5300 commented on modules/mod_manager.ts:
```diff
@@ -142,12 +151,31 @@ class ModManager {
+                    let _url = "";
+                    //TS won't let me delete this bit
+                    //This is a typeguard BTW to separate the collections from the non-collections
+                    if (!(IsInstall(this.source_manager.data))) {
```
I'm unsure if this is the best way to allow an array of the Install type.
It may be better to change all to use an array, and for the pre existing types to get index 0. I think this constant type checking mixed in the current logic is messy.

rob5300 commented on modules/mod_manager.ts:
```diff
@@ -330,22 +397,77 @@ class ModManager {
+                if (!(typeof (this.source_manager.data.setupfunc) == "undefined")) {
+                    setup_func = this.source_manager.data.setupfunc;
+                }
+            }
+        }
+        switch (setup_func) {
+            case "movecfgs":
+                ElectronLog.log("Executing movecfgs install operation")
+                let filesToMove = ["autoexec.cfg", "scout.cfg", "soldier.cfg", "pyro.cfg", "demoman.cfg", "heavyweapons.cfg",
+                "engineer.cfg", "medic.cfg", "sniper.cfg", "spy.cfg"]
+                //Move cfgs
+                let cfgpath = path.join(Main.config.tf2_directory, "tf", "cfg");
+                let usercfgpath = path.join(cfgpath, "user");
+                ElectronLog.log("cfg path is: " + cfgpath);
+                ElectronLog.log("user cfg path is: \"" + usercfgpath + "\"");
+                if (!(fs.existsSync(usercfgpath))) {
+                    //Create usercfgpath
+                    ElectronLog.log("user cfg path does not exist, creating...");
+                    fs.mkdirSync(usercfgpath, {recursive: true});
+                }
+                //Actually move them
+                fs.readdir(cfgpath, (err, files) => {
+                    if (err) {
+                        ElectronLog.error("there was an error on mod_manager.ts, line 447 when trying to read the cfg directory to move cfg files. Trying to read directory: " + cfgpath + " , the error was:" + err)
+                    }
+                    else {
+                        files.forEach(file => {
+                            ElectronLog.log("Checking if should move file: " + file)
+                            if(filesToMove.indexOf(file) > -1) {
+                                ElectronLog.log("Trying to move file " + path.join(cfgpath, file) + " to " + path.join(usercfgpath, path.basename(file)))
+                                fs.rename(path.join(cfgpath ,file), path.join(usercfgpath, path.basename(file)), (err) => {
+                                    if (err) {
+                                        ElectronLog.error("There was an error trying to move file " + path.join(cfgpath, file) + " to " + path.join(usercfgpath, path.basename(file)))
+                                    }
+                                })
+                            }
+                        })
+                    }
+                })
+                break;
```
I like this system to allow configurable post install tasks.
I think a better way is to add a base method to the mod sources, and let that handle these tasks instead of them being done here.
Again i think the many type checks/casts are not good.

How about `ModInstallSource` has a method to handle a PostInstallTask, and this logic can be moved over, and all mod configurations can support this functionality.

rob5300 commented on modules/mod_manager.ts:
```diff
@@ -412,7 +534,55 @@ class ModManager {
+                }
+                else {
+                    if (typeof (this.source_manager.data) != "undefined") {
+                        if (typeof(this.source_manager.data.setupfunc != "undefined")) {
+                            setup_func = this.source_manager.data.setupfunc;  
+                        }
+                    }
+                }
+                switch (setup_func) {
+                    case "movecfgs":
+                        //Move them back
+                        ElectronLog.log("Executing movecfgs uninstall operation")
+                        let filesToMove = ["autoexec.cfg", "scout.cfg", "soldier.cfg", "pyro.cfg", "demoman.cfg", "heavyweapons.cfg",
+                        "engineer.cfg", "medic.cfg", "sniper.cfg", "spy.cfg"]
+                        //Move cfgs
+                        let cfgpath = path.join(Main.config.tf2_directory, "tf", "cfg");
+                        let usercfgpath = path.join(cfgpath, "user");
+                        ElectronLog.log("cfg path is: " + cfgpath);
+                        ElectronLog.log("user cfg path is: \"" + usercfgpath + "\"");
+                        //Actually move them
+                        fs.readdir(usercfgpath, (err, files) => {
+                            if (err)
+                                ElectronLog.error("there was an error on mod_manager.ts, line 566 when trying to read the cfg directory to move cfg files. Trying to read directory: " + cfgpath + "the error was:" + err)
+                            else {
+                                files.forEach(file => {
+                                    ElectronLog.log("Checking if should move file: " + file)
+                                    if(filesToMove.indexOf(file) > -1) {
+                                        ElectronLog.log("Trying to move file " + path.join(usercfgpath, file) + " to " + path.join(cfgpath, path.basename(file)))
+                                        fs.rename(path.join(usercfgpath ,file), path.join(cfgpath, path.basename(file)), (err) => {
+                                            if (err) {
+                                                ElectronLog.error("There was an error trying to move file " + path.join(usercfgpath, file) + " to " + path.join(cfgpath, path.basename(file)))
+                                            }
+                                        })
+                                    }
+                                })
+                            }
+                        })
+                        break;                        
+                    default:
+                        break;
```
Same as above, should be moved also and be similar but named different?

rob5300 commented on modules/mod_sources/collection_source.js:
```diff
@@ -0,0 +1,11 @@
+"use strict";
```
Don't commit your .js files!

rob5300 commented on modules/mod_sources/github_asset.ts:
```diff
@@ -0,0 +1,6 @@
+class Asset {
```
Can this be moved to be with the other ModListEntry related types?

rob5300 commented on modules/configtype.ts:
```diff
@@ -0,0 +1,13 @@
+    current_mod_versions: modVersion[]
+}
+
+class modVersion {
```
Rename to ModVersion

rob5300 reviewed (changes requested):
Left some comments, feel free to reply to ask more.
You can also join our public discord and @rob5300 to better discuss this with me. Really like this so far and would like to get it included when its ready.

If there are changes you are not comfortable doing to other components, I always can.

BernardoGomesNegri commented:
I removed collection_source, now GithubCollectionSource inherits ModInstallSource, most typeguards were removed (but some remain, in mod_manager.ts, lines 164 and 206), changed most functions that used Install to use Install[], renamed ModVersion, moved the Asset class and moved the install/uninstall functions to ModInstallSource.

mastercoms commented on internal/mods.json:
```diff
@@ -102,12 +105,118 @@
             }
+        },
+        {
+            "name": "Mastercomfig",
```
All instances of mastercomfig should be lower case

mastercoms commented on internal/mods.json:
```diff
@@ -102,12 +105,118 @@
+            "instagram": "",
+            "discord": "",
+            "serverlist": "",
+            "contenttext": "Several optimization scripts for modern hardware. Note that after installing, your class cfgs and autoexec.cfg will be moved to a new folder to be compatible with Mastercomfig.",
```
```suggestion
            "contenttext": "A modern Team Fortress 2 performance and customization config. Note that after installing, your class cfgs and autoexec.cfg will be moved to a new folder to be compatible with Mastercomfig.",
```

rob5300 commented on modules/mod_sources/mod_source_base.ts:
```diff
@@ -1,12 +1,112 @@
+                //Actually move them
+                fs.readdir(cfgpath, (err, files) => {
+                    if (err) {
+                        ElectronLog.error("there was an error on mod_manager.ts, line 447 when trying to read the cfg directory to move cfg files. Trying to read directory: " + cfgpath + " , the error was:" + err)
```
I think for these errors, we want to display to the user that something went wrong.
You can use Utilities.ErrorDialog() for this and the isDev package to change the message to a more dev-centric one for development versions.

rob5300 commented on modules/mod_sources/mod_source_base.ts:
```diff
@@ -1,12 +1,112 @@
-    abstract GetFileURL() : Promise<string>;
+    abstract GetFileURL(asset_index? : number) : Promise<string>;
+
+    PostInstall(collection_version?: string) {
```
You are free to make this async if you want.

rob5300 commented on renderer.js:
```diff
@@ -92,6 +94,27 @@ function OnClick_Mod(data) {
+        //Clear the select
+        select.innerHTML = '';
+        //Populate the select
+        let defaultItem
```
Missing semi-colon

rob5300 commented on renderer.js:
```diff
@@ -92,6 +94,27 @@ function OnClick_Mod(data) {
+        });
+    }
+    else {
+        document.getElementById("collection-versions").style.display = "none";
```
If doable, move the element get with the rest.

rob5300 commented on modules/mod_manager.ts:
```diff
@@ -1101,10 +1173,43 @@ function GetFileWriteFunction(extension){
+}
+
+//Find which element in an Install[] has the desiredCollectionString as an argument
+function FindCollectionNumber(Installs: Install[], desiredCollectionString: string) : number {
+    
+    if(typeof(desiredCollectionString) == 'undefined') {
+        return 0
+    }
+
+    if(Installs.length < 2) {
+        //It has a single item
+        return 0
+    }
+    for (let i = 0; i < Installs.length; i++) {
+        const element = Installs[i];
+        if (element.itemname == desiredCollectionString) {
+            //We've found it!
+            return i;
+        }
+    }
+}
```
I just realised this is exported, please add this to Utilities instead, as i wish to change ModManager in future to be instanced per mod.

BernardoGomesNegri commented:
I did make PostInstall async and PostUninstall. however, I do not want to display the "installed" or "uninstalled" dialog before everything is complete. If you wish to not have the "done" dialog wait for moving files, tell me (or commit to the fork)

mastercoms commented on internal/mods.json:
```diff
@@ -102,12 +105,118 @@
+            "backgroundposX": "left",
+            "backgroundposY": "top",
+            "website": "https://mastercomfig.com/",
+            "github": "https://github.com/mastercomfig/mastercomfig",
```
I would rather if only the website was linked, as the website appropriately links to everything else, including the GitHub. I don't really want normal users going on the GitHub as it can be confusing.

rob5300 reviewed (approved):
I approve for this to be merged soon to a branch we made. Then we will update styling.
We will then look to have this in the next release. I plan to make 1/2 changes to help the adding of this feature :)
edit: you are free to merge now/ make a small change if you want and get a re review.
